packages/flutter_tools/test/general.shard/resident_runner_test.dart

```diff
@@ -52,8 +53,18 @@ void main() {
 
   setUp(() {
     testbed = Testbed(setup: () {
+      // Logic to get the filesystem for analytics from globals
+      final ErrorHandlingFileSystem tempFs = globals.fs as ErrorHandlingFileSystem;
+      final MemoryFileSystem analyticsFileSystem;
+      if (tempFs.fileSystem is ThrowingForwardingFileSystem) {
+        analyticsFileSystem =
+            (tempFs.fileSystem as ThrowingForwardingFileSystem).fileSystem
+                as MemoryFileSystem;
+      } else {
+        analyticsFileSystem = tempFs.fileSystem as MemoryFileSystem;
+      }
       fakeAnalytics = getInitializedFakeAnalyticsInstance(
-        fs: globals.fs,
+        fs: analyticsFileSystem,
         fakeFlutterVersion: FakeFlutterVersion(),
```

Comment:
@christopherfujino I'm not the biggest fan of this refactor but we had to consider two cases for this file
1. For most of the tests, we just rely on the `ErrorHandlingFileSystem` in which we pass the `MemoryFileSystem.test` constructor as the delegate, so getting the reference to the filesystem for the analytics constructor was fairly easy
2. There is one test in this file, `HotRunner handles failure to write vmservice file`, which uses `ThrowingForwardingFileSystem` for the delegate in `ErrorHandlingFileSystem`, as a result, I needed to conditionally check for both types before extracting the delegate or I would get the error below
```
type 'ThrowingForwardingFileSystem' is not a subtype of type 'MemoryFileSystem' in type cast
```

Reply:
When there is no override in the test for `FileSystem`
```
ErrorHandlingFileSystem
  --> delegate = MemoryFileSystem
```

When we override the FileSystem in the testUsingContext overrides we have an additional level of nesting
```
ErrorHandlingFileSystem
  --> delegate = ThrowingFileSystem
                              --> delegate = MemoryFileSystem
```

Reply:
Thanks for digging into this. Per my comment [here](https://github.com/flutter/flutter/pull/147250#discussion_r1578330833), I think the best solution would be to just use an implementation of `Analytics` that does not rely on a file system at all. IMO, the file system operations of the `Analytics` class should be tested in `package:unified_analytics`--in tests for package:flutter_tools, all we should be testing is the events that we're sending it.

Reply:
What about for those tests that rely on opt out status, for example, the new [crash reporter refactor](https://github.com/flutter/flutter/pull/147296) to use GA4 opt in status?